### PR TITLE
CB-8807: Refactor cluster upgrade image filtering

### PR DESCRIPTION
**Pull request by tiborpopovics**
In this commit I made the following changes:

- Refactor the filtering logic in ClusterUpgradeImageFilter. The filters are extracted into separate classes. These filters are using a common interface: UpgradeImageFilter
- Refactor the ImageFilterResult class. This contains a List<Image> instead of the Images object.
- Refactor the ImageFilterParams class. This class was extended with the cloudPlatform variable.

### CODE REVIEW

**Review by topolyai5 — CHANGES_REQUESTED** [review]
did you check the optimistic lock in the failed test?
http://ci-cloudbreak.eng.hortonworks.com/job/cloudbreak-pull-request-builder-integration-test/22833/

File: core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/image/filter/CurrentImageUpgradeImageFilter.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+                LOGGER.debug("The current image was removed from the upgrade candidates.");
+                return true;
+            } else {
+                return false;
+            }
+        };
+    }
+
+    private String getReason(List<Image> filteredImages) {
+        return filteredImages.isEmpty() ? "There are no newer compatible images available." : EMPTY_REASON;
```
Comment (lacikaaa) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/image/filter/CurrentImageUpgradeImageFilter.java:55]: this message is ok for `ImageCreationBasedUpgradeImageFilter` but doesn't seem right here

File: core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/image/filter/ImageFilterUpgradeService.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+public class ImageFilterUpgradeService {
+
+    public static final String EMPTY_REASON = "";
+
+    @Inject
+    private Set<UpgradeImageFilter> upgradeImageFilters;
+
+    public ImageFilterResult filterImages(List<Image> availableImages, ImageFilterParams imageFilterParams) {
+        ImageFilterResult result = new ImageFilterResult(availableImages, EMPTY_REASON);
+        for (UpgradeImageFilter imageFilter : upgradeImageFilters) {
```
Comment (lacikaaa) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/image/filter/ImageFilterUpgradeService.java:24]: I have some concerns with this solution:
- it seems to me the filtering order is depending on Spring injection order, which could be different for every startup of the service, at least theoretically, which means 2 filtering could have different results depending which is the last filter. I think the filters should have some ordering, as we had previously.
- we still won't have any idea at least about which images made to the last round of filtering if no images available for upgrade. Most of our questions from customer are like: `Why the upgrade for version X is not possible` I think this refactor should also focus on helping us to answer this question.

**Review by lacikaaa — CHANGES_REQUESTED** [review]

File: core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/image/filter/ImageFilterUpgradeService.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+import javax.inject.Inject;
+
+import org.springframework.stereotype.Service;
+
+import com.sequenceiq.cloudbreak.cloud.model.catalog.Image;
+import com.sequenceiq.cloudbreak.service.upgrade.image.ImageFilterParams;
+import com.sequenceiq.cloudbreak.service.upgrade.image.ImageFilterResult;
+
+@Service
+public class ImageFilterUpgradeService {
```
Comment (lacikaaa) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/image/filter/ImageFilterUpgradeService.java:15]: and also, no unit tests for this one

**Comment by tiborpopovics** [comment]
> I've checked the failed IT and It's not related to my changes. I cannot reproduce in my local environment.

**Comment by topolyai5** [comment]
> > @topolyai5 I've checked the failed IT and It's not related to my changes. I cannot reproduce in my local environment.

What was the root cause?

**Comment by topolyai5** [comment]
> 

Could you please create a jira with the build artifact? We must investigate an optimistic lock to avoid the unnecessary ENGESCs

**Comment by tiborpopovics** [comment]
> > 
> 
> Could you please create a jira with the build artifact? We must investigate an optimistic lock to avoid the unnecessary ENGESCs

Sure, but unfortunately, the artifact is not available anymore. I can only attach the stacktrace.

**Review by topolyai5 — APPROVED** [review]

File: core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/image/filter/CloudPlatformBasedUpgradeImageFilter.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Component;
+
+import com.sequenceiq.cloudbreak.cloud.model.catalog.Image;
+import com.sequenceiq.cloudbreak.service.upgrade.image.ImageFilterParams;
+import com.sequenceiq.cloudbreak.service.upgrade.image.ImageFilterResult;
+
+@Component()
```
Comment (lacikaaa) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/image/filter/CloudPlatformBasedUpgradeImageFilter.java:14]: nit: no need for parenthesis

**Review by lacikaaa — APPROVED** [review]

**Review by lacikaaa — APPROVED** [review]

**Pull request merged**